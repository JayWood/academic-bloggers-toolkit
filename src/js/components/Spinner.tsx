import * as React from 'react';

interface Props {
    /* The length and width in pixels of the spinner*/
    size: string;
    /* Set background color of container */
    bgColor?: string;
    /* If not in a containing element, this is minimum height */
    height?: string | number | (() => string | number);
    /* Add this for a background overlay */
    overlay?: boolean;
    /* Additional style properties for the containing DIV */
    style?: React.CSSProperties;
}

export class Spinner extends React.PureComponent<Props> {
    static defaultProps: Partial<Props> = {
        bgColor: 'transparent',
    };

    style: React.CSSProperties = { ...this.props.style };

    constructor(props: Props) {
        super(props);
        if (this.props.height) {
            if (typeof this.props.height === 'function') {
                this.style.height = this.props.height();
                this.style.minHeight = this.props.height();
            } else {
                this.style.height = this.props.height;
                this.style.minHeight = this.props.height;
            }
        }
        this.style.backgroundColor = this.props.bgColor;
    }

    render() {
        const cn = this.props.overlay
            ? 'abt-spinner abt-spinner_overlay'
            : 'abt-spinner';
        return (
            <div className={cn} style={this.style}>
                <svg
                    width={this.props.size}
                    height={this.props.size}
                    viewBox="0 0 66 66"
                >
                    <circle
                        fill="none"
                        strokeWidth="6"
                        strokeLinecap="round"
                        cx="33"
                        cy="33"
                        r="30"
                    />
                </svg>
                <style jsx>{`
                    @keyframes colors {
                        0% {
                            stroke: #4285f4;
                        }
                        25% {
                            stroke: #de3e35;
                        }
                        50% {
                            stroke: #f7c223;
                        }
                        75% {
                            stroke: #1b9a59;
                        }
                        100% {
                            stroke: #4285f4;
                        }
                    }
                    @keyframes dash {
                        0% {
                            stroke-dashoffset: 187;
                        }
                        50% {
                            stroke-dashoffset: 46.75;
                            -webkit-transform: rotate(135deg);
                            transform: rotate(135deg);
                        }
                        100% {
                            stroke-dashoffset: 187;
                            -webkit-transform: rotate(450deg);
                            transform: rotate(450deg);
                        }
                    }
                    @keyframes rotator {
                        0% {
                            -webkit-transform: rotate(0deg);
                            transform: rotate(0deg);
                        }
                        100% {
                            -webkit-transform: rotate(270deg);
                            transform: rotate(270deg);
                        }
                    }
                    .abt-spinner {
                        display: flex;
                        height: 100%;
                        width: 100%;
                        pointer-events: none;
                        align-items: center;
                        justify-content: center;
                    }
                    .abt-spinner_overlay {
                        position: absolute;
                        background: rgba(0, 0, 0, 0.2);
                    }
                    svg {
                        animation: rotator 1.4s linear infinite;
                    }
                    circle {
                        stroke-dasharray: 187;
                        stroke-dashoffset: 0;
                        transform-origin: center;
                        animation: dash 1.4s ease-in-out infinite,
                            colors 5.6s ease-in-out infinite;
                    }
                `}</style>
            </div>
        );
    }
}
