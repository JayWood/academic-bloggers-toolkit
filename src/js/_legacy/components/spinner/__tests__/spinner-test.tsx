import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import React from 'react';

import Spinner from '..';

describe('<Spinner />', () => {
    it('should match snapshots', () => {
        let component = shallow(<Spinner size="50px" />);
        expect(toJSON(component)).toMatchSnapshot();

        component = shallow(
            <Spinner size="100px" bgColor="magenta" height={100} overlay />,
        );
        expect(toJSON(component)).toMatchSnapshot();

        const heightFn = () => '100px';
        component = shallow(<Spinner size="100px" height={heightFn} />);
        expect(toJSON(component)).toMatchSnapshot();
    });
});
