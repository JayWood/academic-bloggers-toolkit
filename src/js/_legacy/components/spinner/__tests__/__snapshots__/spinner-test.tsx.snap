// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Spinner /> should match snapshots 1`] = `
<div
  className="spinner "
  style={
    Object {
      "backgroundColor": "transparent",
    }
  }
>
  <svg
    height="50px"
    viewBox="0 0 66 66"
    width="50px"
  >
    <circle
      cx="33"
      cy="33"
      fill="none"
      r="30"
      strokeLinecap="round"
      strokeWidth="6"
    />
  </svg>
</div>
`;

exports[`<Spinner /> should match snapshots 2`] = `
<div
  className="spinner spinnerOverlay"
  style={
    Object {
      "backgroundColor": undefined,
      "height": 100,
      "minHeight": 100,
    }
  }
>
  <svg
    height="100px"
    viewBox="0 0 66 66"
    width="100px"
  >
    <circle
      cx="33"
      cy="33"
      fill="none"
      r="30"
      strokeLinecap="round"
      strokeWidth="6"
    />
  </svg>
</div>
`;

exports[`<Spinner /> should match snapshots 3`] = `
<div
  className="spinner "
  style={
    Object {
      "backgroundColor": "transparent",
      "height": "100px",
      "minHeight": "100px",
    }
  }
>
  <svg
    height="100px"
    viewBox="0 0 66 66"
    width="100px"
  >
    <circle
      cx="33"
      cy="33"
      fill="none"
      r="30"
      strokeLinecap="round"
      strokeWidth="6"
    />
  </svg>
</div>
`;
